### ISSUE TITLE
Invalid JSON-LD if Jena defines an "" prefix

### ISSUE THREAD

**Issue by fpservant** [jsonld-java/jsonld-java#141]
if a prefix is set to "", we get an output that is not correct JSON-LD
(according to the JSON-LD playground: "Invalid JSON-LD syntax; a term cannot be an empty string.")

Here a small program using Jena that illustrates the problem. It creates a model and outputs it in JSON-LD. Jena's output in turtle is :

@prefix :      http://www.a.com/foo/ .

:s      :p      :o .

package sometests;
import org.junit.Test;
import com.hp.hpl.jena.rdf.model.Model;
import com.hp.hpl.jena.rdf.model.ModelFactory;
import com.hp.hpl.jena.rdf.model.Property;
import com.hp.hpl.jena.rdf.model.Resource;

public class JsonLDWriterTest {

@Test
public final void test3() {
    Model m = ModelFactory.createDefaultModel();
    String url = "http://www.a.com/foo/";
    Resource s = m.createResource(url + "s");
    Property p = m.createProperty(url + "p");
    Resource o = m.createResource(url + "o");
    m.add(s,p,o);
    m.setNsPrefix("",url);
    m.write(System.out, "JSON-LD");
}

}

**Comment by ansell** [comment]
The only direction that the spec appears to give in this case is in 6.7:

http://www.w3.org/TR/json-ld/#advanced-context-usage

Furthermore, the use of empty terms ("") is not allowed as not all programming languages are able to handle empty JSON keys.

The Jena package is also not maintained here, although it may be a bug in the core not to throw a more major error. @afs What is your interpretation of this knowing the Jena implementation and the text from the spec?

**Comment by ansell** [comment]
Sorry, 6.7 is a non-normative reference, the normative text is in 8.1, but that isn't referenced from 6.7 for some reason:

http://www.w3.org/TR/json-ld/#h3_terms

    Furthermore, the term MUST NOT be an empty string ("") as not all programming languages are able to handle empty JSON keys.

**Comment by afs** [comment]
@ansell - yes, this is hard to follow in the spec.8.1 is indirectly referring to JSON keys because a key becomes a URI.  But by the same text, it applies to values as well (they can be expanded to URIs) and that makes "Example 15" violate that text.

I think @fpservant is right; the JSON-LD playground does indeed give a error on the JSON-LD produces.

[JENA-934](https://issues.apache.org/jira/browse/JENA-934) fixes this by not setting "" as a prefix passed to jsonld-java, merely skipping it.  That at least produces output acceptable to the JSON-LD playground.
